mllib/src/main/scala/org/apache/spark/ml/feature/OneHotEncoderEstimator.scala

```diff
@@ -0,0 +1,464 @@
+      s"The number of input columns ${inputColNames.length} must be the same as the number of " +
+        s"output columns ${outputColNames.length}.")
+
+    require(inputColNames.length == categorySizes.length,
+      s"The number of input columns ${inputColNames.length} must be the same as the number of " +
+        s"features ${categorySizes.length} during fitting.")
+
+    val inputOutputPairs = inputColNames.zip(outputColNames)
+    val outputFields = inputOutputPairs.map { case (inputColName, outputColName) =>
+
+      require(schema(inputColName).dataType.isInstanceOf[NumericType],
+        s"Input column must be of type NumericType but got ${schema(inputColName).dataType}")
+      require(!inputFields.exists(_.name == outputColName),
+        s"Output column $outputColName already exists.")
+
+      OneHotEncoderCommon.transformOutputColumnSchema(
+        schema(inputColName), $(dropLast), outputColName)
+    }
+    verifyNumOfValues(StructType(inputFields ++ outputFields))
+  }
+
+  private def verifyNumOfValues(schema: StructType): StructType = {
+    $(outputCols).zipWithIndex.foreach { case (outputColName, idx) =>
+      val inputColName = $(inputCols)(idx)
+      val attrGroup = AttributeGroup.fromStructField(schema(outputColName))
+
+      // If the input metadata specifies number of category,
+      // compare with expected category number.
+      if (attrGroup.attributes.nonEmpty) {
+        require(attrGroup.size == categorySizes(idx), "OneHotEncoderModel expected " +
+          s"${categorySizes(idx)} categorical values for input column ${inputColName}, but " +
+            s"the input column had metadata specifying ${attrGroup.size} values.")
+      }
+    }
+    schema
+  }
+
+  @Since("2.3.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    if (getDropLast && getHandleInvalid == OneHotEncoderEstimator.SKIP_INVALID) {
+      throw new IllegalArgumentException("When Param handleInvalid is set to " +
+        s"${OneHotEncoderEstimator.SKIP_INVALID}, Param dropLast can't be true, " +
+        "because last category and invalid values will conflict in encoded vector.")
+    }
+
+    val transformedSchema = transformSchema(dataset.schema, logging = true)
+
+    val encodedColumns = encoders.zipWithIndex.map { case (encoder, idx) =>
+      val inputColName = $(inputCols)(idx)
+      val outputColName = $(outputCols)(idx)
+
+      val outputAttrGroupFromSchema =
+        AttributeGroup.fromStructField(transformedSchema(outputColName))
+
+      val metadata = if (outputAttrGroupFromSchema.size < 0) {
+        OneHotEncoderCommon.createAttrGroupForAttrNames(outputColName, false,
+          categorySizes(idx)).toMetadata()
+      } else {
+        outputAttrGroupFromSchema.toMetadata()
+      }
+
+      encoder(col(inputColName).cast(DoubleType)).as(outputColName, metadata)
+    }
+    val allCols = Seq(col("*")) ++ encodedColumns
+    dataset.select(allCols: _*)
```

Comment:
Does we need to handle the case: `outputColNames` conflict with existing column names ?

Reply:
Yeah my other https://github.com/apache/spark/pull/19527#discussion_r145455661 was that we created `withColumns` in `Dataset` that handles this? https://github.com/apache/spark/blob/fa225da7463e384529da14706e44f4a09772e5c1/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L2092

Reply:
@WeichenXu123 Yes. `transformSchema` has handled the case that there are existing columns which conflict with any of `outputColNames`. 

@MLnick Sorry I misunderstood your comment https://github.com/apache/spark/pull/19527#discussion_r145455661. `withColumns` approach calls `select` so there is no performance difference. `withColumns` did one more thing to detect duplication in the column names. We should check the duplication too.

Reply:
@MLnick Use `withColumns` now.